tools/discovery/discovery.go

```diff
@@ -0,0 +1,131 @@
+package discovery
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+
+	"github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+	"github.com/xiekeyang/oci-discovery/tools/object"
+)
+
+var (
+	defaultTrans = &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+)
+
+func DiscoveryHandler(context *cli.Context) error {
+	var (
+		name  = context.Args()[0]
+		roots []v1.Descriptor
+	)
+
+	v, err := paramsParser(name)
+	if err != nil {
+		return err
+	}
+
+	engines, err := refEnginesFetching(v)
+	if err != nil {
+		return err
+	}
+
+	for _, engine := range engines.RefEngines {
+		var ur urlResolver = urlResolver(engine.Uri)
+		u, err := ur.resolve(v)
```

Comment:
This assumes that the `protocol` is always `oci-index-template-v1`.  I'd rather have a ref-engine interface (a trimmed down and adjusted form of [this][1]):

```go
type Engine interface {
  Get(ctx context.Context, name string) (descriptors []specs.Descriptor, err error)
  Close() (err error)
}
```

A ref-engine generation template:

```go
type NewEngine func(ctx context.Context, config interface{}) (engine Engine, err error)
```

and a [registry of per-protocol generators][2]:

```go
var Constructors map[string]NewEngine {
  "oci-index-template-v1": NewIndexTemplateEngine,
  …
}
```
[1]: https://github.com/opencontainers/image-tools/pull/5/files#diff-d8af84c58f6ae3369e0b5cbca06727dfR27
[2]: https://github.com/xiekeyang/oci-discovery/blame/44ec3cf3113e29a743ad04220ccb7ff5197dab2a/oci_discovery/ref_engine/__init__.py#L19-L22

Reply:
> I'd rather have a ref-engine interface…

Floated with #23.